server/sonar-server/src/main/java/org/sonar/server/webhook/WebhookPayloadFactoryImpl.java

```diff
@@ -0,0 +1,161 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.webhook;
+
+import java.io.StringWriter;
+import java.io.UnsupportedEncodingException;
+import java.io.Writer;
+import java.net.URLEncoder;
+import java.util.Date;
+import java.util.Map;
+import org.sonar.api.ce.ComputeEngineSide;
+import org.sonar.api.platform.Server;
+import org.sonar.api.utils.System2;
+import org.sonar.api.utils.text.JsonWriter;
+
+import static java.lang.String.format;
+import static org.sonar.core.config.WebhookProperties.ANALYSIS_PROPERTY_PREFIX;
+
+@ComputeEngineSide
+public class WebhookPayloadFactoryImpl implements WebhookPayloadFactory {
+
+  private final Server server;
+  private final System2 system2;
+
+  public WebhookPayloadFactoryImpl(Server server, System2 system2) {
+    this.server = server;
+    this.system2 = system2;
+  }
+
+  @Override
+  public WebhookPayload create(ProjectAnalysis analysis) {
+    Writer string = new StringWriter();
+    try (JsonWriter writer = JsonWriter.of(string)) {
+      writer.beginObject();
+      writeServer(writer);
+      writeTask(writer, analysis.getCeTask());
+      writeDates(writer, analysis, system2);
+      writeProject(analysis, writer, analysis.getProject());
+      analysis.getBranch().ifPresent(b -> writeBranch(writer, analysis.getProject(), b));
+      analysis.getQualityGate().ifPresent(qualityGate -> writeQualityGate(writer, qualityGate));
+      writeAnalysisProperties(writer, analysis.getProperties());
+      writer.endObject().close();
+      return new WebhookPayload(analysis.getProject().getKey(), string.toString());
+    }
+  }
+
+  private void writeServer(JsonWriter writer) {
+    writer.prop("serverUrl", server.getPublicRootUrl());
+  }
+
+  private static void writeDates(JsonWriter writer, ProjectAnalysis analysis, System2 system2) {
+    analysis.getAnalysisDate().ifPresent(date -> writer.propDateTime("analysedAt", date));
+    writer.propDateTime("changedAt", analysis.getAnalysisDate().orElseGet(() -> new Date(system2.now())));
+  }
+
+  private void writeProject(ProjectAnalysis analysis, JsonWriter writer, Project project) {
+    writer
+      .name("project")
+      .beginObject()
+      .prop("key", project.getKey())
+      .prop("name", analysis.getProject().getName())
+      .prop("url", projectUrlOf(project))
+      .endObject();
+  }
+
+  private static void writeAnalysisProperties(JsonWriter writer, Map<String, String> properties) {
+    writer
+      .name("properties")
+      .beginObject();
+    properties.entrySet()
+      .stream()
+      .filter(prop -> prop.getKey().startsWith(ANALYSIS_PROPERTY_PREFIX))
+      .forEach(prop -> writer.prop(prop.getKey(), prop.getValue()));
+    writer.endObject();
+  }
+
+  private static void writeTask(JsonWriter writer, CeTask ceTask) {
+    writer
+      .prop("taskId", ceTask.getId())
+      .prop("status", ceTask.getStatus().toString());
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "status" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1192)